### PR TITLE
User profile picture implementation

### LINKED ISSUE #6: Update user profile picture
- [x] Users should be able to see their current profile picture at the page. If they didn't upload the picture yet, then they should see a placeholder.
- [x] Users should be able to update their profile picture.
- [x] Users should not be able to upload images which size is more than 1024x1024px

### PR DESCRIPTION
**Pull request by Washubrnfg85**

### CODE REVIEW

File: src/main/java/org/linksharing/server/http/ApiRestController.java [hunk 1]
```diff
@@ -67,7 +67,16 @@ ResponseEntity<UserProfileDetails> updateProfileDetails(Principal user, @Request
 
     @GetMapping("/profile/picture")
     ResponseEntity<?> getProfilePicture(Principal user) throws FileNotFoundException {
-        File img = new File("../image.jpg");
+
+        String imageUrl;
+
+        try {
+            imageUrl = profileRepository.findByEmail(user.getName()).getImageUrl();
+        } catch (NullPointerException e) {
```
Comment (eugenesmkn) [inline src/main/java/org/linksharing/server/http/ApiRestController.java]: Usually, we don't catch NPE. The better way is check is something is `null` or not.

Comment (Washubrnfg85) [reply src/main/java/org/linksharing/server/http/ApiRestController.java]: Ok. Fixed.

**Review by eugenesimakin — COMMENTED** [review]
Looks good

File: src/main/java/org/linksharing/server/http/ApiRestController.java [hunk 2]
```diff
@@ -66,15 +68,15 @@ ResponseEntity<UserProfileDetails> updateProfileDetails(Principal user, @Request
...
+    ResponseEntity<?> getProfilePicture(Principal user) throws IOException {
 
         UserProfileDetails userProfile = profileRepository.findByEmail(user.getName());
         String imageUrl;
 
         if (userProfile.getImageUrl() != null) {
             imageUrl = userProfile.getImageUrl();
         } else  {
-            imageUrl = "../image.jpg";
+            imageUrl = "src/main/resources/static/placeholder.jpg";
```
Comment (eugenesmkn) [inline src/main/java/org/linksharing/server/http/ApiRestController.java:84]: Have you tried to run the app as standalone jar file and check if this works?

**Comment by eugenesmkn** [comment]
Checkstyle configuration added. Rebase to main.

File: src/main/java/org/linksharing/server/http/ApiRestController.java (hunk 2 above)
Comment (Washubrnfg85) [reply src/main/java/org/linksharing/server/http/ApiRestController.java:84]: Nope. But if you ask then I suppose it won't.
I didn't know what path is correct and calculated that one like this:
`new File("").toURI().relativize(new File("D:/Java_Projects/link-sharing-app/src/main/resources/static/placeholder.jpg").toURI()).getPath();`
It should be relative but looks like absolut which is a little odd. 
Anyway it works at the moment, so I put it on hold for now.

**Comment by Washubrnfg85** [comment]
Oh, crap!! Didn't mention your last commit. Will fix it tomorrow.

**Comment by Washubrnfg85** [comment]
I don't know how to fix that check failure.

**Comment by eugenesmkn** [comment]
> I don't know how to fix that check failure.

You can see the error here https://github.com/eugenesimakin/link-sharing-app/actions/runs/7314679783/job/19927550955?pr=14 
or you can build the project locally using gradle and see the same error.

**Comment by eugenesmkn** [comment]
Looks like there is a bug in saving profile details (not profile picture). I added some tests so try to rebase to the main branch and run the tests. Or try to find the bug manually.
Note: For running tests locally, remove "-headless" option.

File: src/main/java/org/linksharing/server/http/ApiRestController.java (hunk 2 above)
Comment (eugenesmkn) [reply src/main/java/org/linksharing/server/http/ApiRestController.java:84]: When running jar file on server you will get this:
> java.io.FileNotFoundException: src\main\resources\static\placeholder.jpg

Use `org.springframework.core.io.ClassPathResource` with correct path within class path.

File: src/main/java/org/linksharing/server/http/ApiRestController.java [hunk 3]
```diff
@@ -66,18 +70,39 @@ ResponseEntity<UserProfileDetails> updateProfileDetails(Principal user, @Request
...
     @PostMapping("/profile/picture")
-    ResponseEntity<?> updateProfilePicture(Principal user) {
-        return null;
+    ResponseEntity<UserProfileDetails> updateProfilePicture(Principal user, @RequestParam("file") MultipartFile file) throws IOException {
+        UserProfileDetails userProfile = profileRepository.findByEmail(user.getName());
+
+        String[] fileNameParts = file.getOriginalFilename().split("\\.");
+        String fileExtension = "." + fileNameParts[fileNameParts.length - 1];
+
+        Path newFileName = Paths.get("src/main/resources/upload/", userProfile.getEmail() + fileExtension);
```
Comment (eugenesmkn) [inline src/main/java/org/linksharing/server/http/ApiRestController.java]: Choose some folder which is in the same folder as the working dir (working directory is the directory where the app is running). 
If you're running the app from Idea the working dir is equal to project dir. 
Don't forget to exclude the folder from Git. We don't need user data in here.

**Comment by Washubrnfg85** [comment]
> You can see the error here https://github.com/eugenesimakin/link-sharing-app/actions/runs/7314679783/job/19927550955?pr=14
or you can build the project locally using gradle and see the same error.

Yeah, but I removed that line 93. And recommited it, but still...

File: src/main/java/org/linksharing/server/http/ApiRestController.java (hunk 3 above)
Comment (Washubrnfg85) [reply src/main/java/org/linksharing/server/http/ApiRestController.java]: To go after New Year)))

**Comment by eugenesmkn** [comment]
> Yeah, but I removed that line 93. And recommited it, but still...

The error message is 
`[ERROR] /home/runner/work/link-sharing-app/link-sharing-app/src/main/java/org/linksharing/server/http/ApiRestController.java:93: Line is longer than 120 characters (found 138). [LineLength]`

Here is the key part `Line is longer than 120 characters (found 138)`.

PS. Do you want to get rid of checkstyle? It's not as good as I thought and not as good as linters for javascript in vscode.

**Comment by Washubrnfg85** [comment]
> > Yeah, but I removed that line 93. And recommited it, but still...
> 
> The error message is `[ERROR] /home/runner/work/link-sharing-app/link-sharing-app/src/main/java/org/linksharing/server/http/ApiRestController.java:93: Line is longer than 120 characters (found 138). [LineLength]`
> 
> Here is the key part `Line is longer than 120 characters (found 138)`.
> 
I know. Tried to fix it. Maybe was tired.

> PS. Do you want to get rid of checkstyle? It's not as good as I thought and not as good as linters for javascript in vscode.
>
Yes, please.

**Comment by eugenesmkn** [comment]
> > PS. Do you want to get rid of checkstyle? It's not as good as I thought and not as good as linters for javascript in vscode.
> 
> Yes, please.

Done. Rebase to main.

File: src/main/java/org/linksharing/server/http/ApiRestController.java (hunk 3 above)
Comment (Washubrnfg85) [reply src/main/java/org/linksharing/server/http/ApiRestController.java]: You mean that the "pics" folder should be just under the working directory?
Like this?
MyProjects -> link-sharing-app -> pics

Or next to working directory, like this?
MyProjects -> link-sharing-app
MyProjects -> pics
MyProjects -> some-other-project

**Comment by Washubrnfg85** [comment]
Please check out last commit. Something wrong with profile-details-page.html (script perhaps). Save button is disabled. Can't figure out why.

File: src/main/java/org/linksharing/server/http/ApiRestController.java (hunk 3 above)
Comment (eugenesmkn) [reply src/main/java/org/linksharing/server/http/ApiRestController.java]: > MyProjects -> link-sharing-app -> pics

Look good to me.

The idea is that when application starts (from jar or from Intellij Idea), it creates the "pics" directory in the same directory where it's running and then uses the "pics" directory for storing profile pictures.

File: src/main/resources/templates/profile_details_page.html [hunk 4]
```diff
@@ -112,8 +112,43 @@ <h4 class="alert-heading">Error while loading profile details</h4>
                 firstName: '',
                 lastName: '',
                 publicEmail: '',
+                file: null,
             });
 
+            function updateStateFromResponse(resp) {
```
Comment (eugenesmkn) [inline src/main/resources/templates/profile_details_page.html:118]: This function is already exists

**Pull request closed without merging**